cpp/src/arrow/io/compressed.cc

```diff
@@ -284,8 +308,12 @@ class CompressedInputStream::Impl {
     int64_t decompress_size = kDecompressSize;
 
     while (true) {
-      ARROW_ASSIGN_OR_RAISE(decompressed_,
-                            AllocateResizableBuffer(decompress_size, pool_));
+      if (decompressed_ == nullptr) {
+        ARROW_ASSIGN_OR_RAISE(decompressed_,
+                              AllocateResizableBuffer(decompress_size, pool_));
+      } else {
+        RETURN_NOT_OK(decompressed_->Resize(decompress_size, /*shrink_to_fit=*/false));
```

Comment:
Reallocating like this will need to copy the existing data, but we don't need the data. It is not obvious it is faster than allocating a new buffer, it probably depends on the underlying allocator.

@felipecrv Thoughts?

Reply:
I'll note that mimalloc has [mi_expand](https://microsoft.github.io/mimalloc/group__malloc.html#gaaee66a1d483c3e28f585525fb96707e4) and jemalloc has [xallocx](https://jemalloc.net/jemalloc.3.html) which could be used to implement a resize-in-place API at the MemoryPool level. This would be useful in many instances of resizing a scratch space without caring about the contents.

Reply:
Opened https://github.com/apache/arrow/issues/40686 for this.

Reply:
> Mostly I think actually no really Resize happens, the size is likely to be same as `kDecompressSize`

That's quite possible. You could try to log this and see what happens during the tests?

Reply:
Oh mybad, I'm wrong here. `decompressed_` is likely to have has same capacity, but not same-sized

Reply:
Sorry I wrote this a few days ago, I made it wrong. Applogize for this. `decompressed_` would `Resize` when final output, which could make it smaller. This edit to `decompressed_` might not bring optimization except for some weird case